sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java

```diff
@@ -406,11 +406,25 @@ public VisitDecision preorder(DBSPSourceMapOperator operator) {
         this.generateFromTrait(type);
         // Macro for serialization
         this.generateRenameMacro(operator.outputName, type, operator.metadata);
+
+        DBSPTypeStruct keyStructType = operator.getKeyStructType(
+                // TODO: this should be a fresh name.
+                operator.originalRowType.sanitizedName + "_key");
         // Generate key type definition
-        item = new DBSPStructItem(operator.getKeyStructType());
+        item = new DBSPStructItem(keyStructType);
         item.accept(this.innerVisitor);
+
         // Trait for serialization
-        this.generateFromTrait(operator.getKeyStructType());
+        this.generateFromTrait(keyStructType);
+        this.generateRenameMacro(item.type.name, item.type, operator.metadata);
+
+        DBSPTypeStruct upsertStruct = operator.getStructUpsertType(
+                operator.originalRowType.sanitizedName + "_upsert");
```

Comment:
Should we be worried about name clashes?

Reply:
Yes